apps/nn_ops/MaxPool_generator.cpp

```diff
@@ -0,0 +1,103 @@
+// 'MaxPool' implementation in Halide.
+
+#include "common.h"
+#include <Halide.h>
+
+using Halide::BoundaryConditions::constant_exterior;
+using Halide::ConciseCasts::u8_sat;
+using Halide::Generator;
+
+class MaxPool : public Generator<MaxPool> {
+ public:
+  // Unsigned 8-bit input tensor, indexed by depth, x, y, batch.
+  ImageParam input_{UInt(8), 4, "input"};
```

Comment:
You should generally prefer using Input<> (rather than ImageParam and Param<>) for new Generators.

Reply:
This might be matching what I did in MatrixMultiply, in which case, I'd say let's just keep matching that, and we can deal with this later all at once for these generators (probably still before we submit to master - note this PR is to the nn-ops branch, not master).

Reply:
Yep, I can change them all together later.